packages/drivers/driver-base/src/documentDeltaConnection.ts

```diff
@@ -149,11 +159,31 @@ export class DocumentDeltaConnection
             // though some logic (naming assert in initialMessages getter) might need to be adjusted (it becomes noop)
             assert((this.listeners(event).length !== 0) === this.trackedListeners.has(event), 0x20b /* "mismatch" */);
             if (!this.trackedListeners.has(event)) {
-                this.addTrackedListener(
-                    event,
-                    (...args: any[]) => {
-                        this.emit(event, ...args);
-                    });
+                let listener: (...args: any[]) => void;
+                switch (event) {
+                    case "op": {
+                        listener = (docId: string, messages: ISequencedDocumentMessage[]) => {
+                            if (!this.lightOpAcksEnabled) {
+                                this.emit(event, docId, messages);
+                            } else {
+                                this.emit(event, docId, messages.map((message) => {
+                                    // Repopulate op contents if applicable
+                                    if (this.clientId === message.clientId && this.opContent.has(message.clientSequenceNumber)) {
+                                        message.contents = this.opContent.get(message.clientSequenceNumber);
```

Comment:
Also, if there is such opportunity - it would be nice to assert that we are not leaking memory by accumulating (and not removing) entries form this.opContent. One place where it actually happens (and possibly pointing to us changing design) is noops and coalescing by service - client sends noops, but service may never sequence them (and thus skip localSequenceNumber). Over time we might accumulate a lot of them.

That all said, ops on the wire can come out of order (I think, at least we do not make any assumptions about order anywhere in the stack). If that's the case, cleaning this.opContent becomes very hard (as there no definitive signal or event that we can use to do so) Maybe we disregard my suggestion on being defensive here (prev comment) and simply not remember ops with empty content? And at the time of disconnect simply verify that we have no more than X entries in this.opContent? (X would need to be vey large, unfortunately, so usefulness of this check would be very low).

Let's brainstorm it together, I have no good suggestion here other than raising the problem